src/collective/listmonk/services/subscriptions.py

```diff
@@ -0,0 +1,216 @@
+    list_ids: list[int]
+    email: str
+
+
+class Unsubscribe(PydanticService):
+    def reply(self):
+        data = self.validate(UnsubscribeRequest)
+        subscriber = get_subscriber(data.email)
+        if subscriber is None:
+            raise BadRequest("Subscription not found")
+        call_listmonk(
+            "put",
+            "/subscribers/lists",
+            json={
+                "ids": [subscriber["id"]],
+                "action": "unsubscribe",
+                "target_list_ids": data.list_ids,
+            },
+        )
+
+
+# TODO get real values from configuration
+listmonk = RelativeSession("http://localhost:9000/api")
+listmonk.auth = ("admin", "admin")
+
+
+def call_listmonk(method, path, **kw):
+    func = getattr(listmonk, method.lower())
+    response = func(path, **kw)
+    response.raise_for_status()
+    return response.json()
+
+
+def get_subscriber(email: str) -> Optional[dict]:
+    result = call_listmonk(
+        "get",
+        "/subscribers",
+        params={"query": f"email='{email}'"},
+    )
+    count = result["data"]["total"]
+    if count == 1:
+        return result["data"]["results"][0]
+    elif count > 1:
+        raise BadRequest("Found more than one subscriber")
+
+
+def get_pending_confirmation_storage() -> OOBTree:
+    """Get or create the BTree used to track pending confirmations."""
+    portal = api.portal.get()
+    if not hasattr(portal, "_listmonk_pending_confirmations"):
+        portal._listmonk_pending_confirmations = OOBTree()
+    return portal._listmonk_pending_confirmations
+
+
+def create_pending_confirmation(
+    sub_id: int, data: SubscriptionRequest
+) -> PendingConfirmation:
+    storage = get_pending_confirmation_storage()
+    token = uuid.uuid4().hex
+    pc = PendingConfirmation(token=token, sub_id=sub_id, list_ids=data.list_ids)
+    storage[token] = pc.model_dump()
+    return pc
+
+
+def send_confirmation(data: SubscriptionRequest, confirm_link: str):
+    mailhost = api.portal.get_tool("MailHost")
+    # lang = api.portal.get_current_language()
+    # TODO show list titles instead of ids
+    # TODO DE translation
+    subject = "Confirm Subscription"
+    body = f"""Confirm subscription
+
+Someone has requested a subscription to the following newsletters: {data.list_ids}
+
+To confirm this subscription, click this link:
+{confirm_link}
+
+If you did not request this subscription, you can ignore this email.
+"""
+    message = EmailMessage()
+    message.set_content(body)
+    mail_settings = get_mail_settings()
+    from_name = mail_settings.email_from_name
+    from_email = mail_settings.email_from_address
+    mailhost.send(
+        message.as_bytes(),
+        data.email,
+        formataddr((from_name, from_email)),
+        subject=subject,
+        charset="utf-8",
+        immediate=True,
+    )
+
+
+def get_mail_settings():
+    registry = getUtility(IRegistry)
+    return registry.forInterface(IMailSchema, prefix="plone")
```

Comment:
why not use api.portal.get_registry?

Reply:
@ericof I didn't know about it! I'll take a look

Reply:
@ericof `get_registry_setting` in plone.api is a bit awkward and less efficient when I want to access more than one setting from the same schema, like here.